arangod/Cluster/HeartbeatThread.cpp

```diff
@@ -374,17 +418,26 @@ void HeartbeatThread::getNewsFromAgencyForDBServer() {
     }
   }
 
-  // Periodically update the list of DBServers and prune agency comm
-  // connection pool:
-  if (++_updateCounter >= 60) {
+  // Periodical or event-based update of DBServers and pruning of
+  // agency comm connection pool:
+  if (_updateDBServers.load() || ++_updateCounter >= 60) {
     auto& clusterFeature = server().getFeature<ClusterFeature>();
     auto& ci = clusterFeature.clusterInfo();
     ci.loadCurrentDBServers();
     _updateCounter = 0;
+    _updateDBServers.store(false);
```

Comment:
You could miss an update that happened after the load this way. To avoid that, you could do a compare-exchange instead of the load, and omit the store.